File: completable-reactor-graph-viewer/src/test/java/ru/fix/completable/reactor/graph/viewer/GraphViewerTest.java
package ru.fix.completable.reactor.graph.viewer;

import javafx.application.Application;
import javafx.scene.input.KeyCode;
import javafx.stage.Stage;
import lombok.val;
import org.apache.commons.io.IOUtils;
import org.junit.Ignore;
import org.junit.Test;
import ru.fix.completable.reactor.api.ReactorGraphModel;
import ru.fix.completable.reactor.graph.viewer.code.CodeUpdater;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by swarmshine on 29.01.2017.
 */

public class GraphViewerTest extends Application {

    private static String payload;

    public GraphViewerTest() {
    }

    static class LogActionListener implements GraphViewer.ActionListener {
        @Override
        public void goToSource(ReactorGraphModel.Source source) {
            System.out.println("goToSource: " + source);
        }

        @Override
        public void goToSubgraph(String subgraphPayloadClass) {
            System.out.println("goToSubgraph: " + subgraphPayloadClass);
        }

        @Override
        public void coordinatesChanged(List<GraphViewer.CoordinateItem> coordinateItems) {
            try {
                System.out.println("changeCoordinates:\n" +
                        coordinateItems.stream().map(Object::toString).collect(Collectors.joining("\n")));

                val codeUpdater = new CodeUpdater();
                try (val resource = getClass().getResourceAsStream("/viewer-code-block.txt")) {
                    System.out.println("Code:\n" +
                            codeUpdater.updateCoordinates(IOUtils.toString(resource, StandardCharsets.UTF_8), coordinateItems));
                }
            } catch (Exception exc) {
                exc.printStackTrace();
            }
        }
    }

    @Override
    public void start(Stage stage) throws Exception {
        if (stage == null) {
            throw new IllegalArgumentException("stage is null");
        }


        String graphModel;

        try (val resource = getClass().getResourceAsStream("/"  + GraphViewerTest.payload)) {

            graphModel = IOUtils.toString(resource, StandardCharsets.UTF_8);
        }

        val viewer = new GraphViewer();
        viewer.registerListener(new LogActionListener());

        viewer.setShortcut(ShortcutType.GOTO_SERIALIZATION_POINT, new Shortcut(true, KeyCode.B));

        viewer.openGraph(graphModel);
        stage.setScene(viewer.getScene());

        stage.show();
    }

    /**
     * Manual UI Test
     */
    @Ignore
    @Test
    public void open_pane() throws Exception {
        GraphViewerTest.payload = "PurchasePayload.rg";
        Application.launch();
    }


    @Ignore
    @Test
    public void open_grpah_with_subgraph() throws Exception{
        GraphViewerTest.payload = "SubscribePayload.rg";
        Application.launch();

    }

}

File: completable-reactor-graph-viewer-app/src/test/kotlin/ru/fix/completable/reactor/graph/viewer/app/GraphViewerTest.kt
package ru.fix.completable.reactor.graph.viewer.app

import javafx.application.Application
import javafx.scene.input.KeyCode
import javafx.stage.Stage
import mu.KotlinLogging
import org.junit.Ignore
import org.junit.Test
import ru.fix.completable.reactor.api.gl.model.Source
import ru.fix.completable.reactor.graph.viewer.gl.CoordinateItem
import ru.fix.completable.reactor.graph.viewer.gl.GraphViewer
import ru.fix.completable.reactor.graph.viewer.gl.Shortcut
import ru.fix.completable.reactor.graph.viewer.gl.ShortcutType
import ru.fix.completable.reactor.parser.java.JavaSourceParser


private val log = KotlinLogging.logger {}

/**
 * Created by swarmshine on 29.01.2017.
 */

class GraphViewerTest : Application() {

    companion object {
        private var payload: String? = null
    }


    class LogActionListener : GraphViewer.ActionListener {
        override fun goToSource(source: Source) {
            println("goToSource: " + source)
        }

        override fun goToSubgraph(subgraphPayloadType: String) {
            println("goToSubgraph: " + subgraphPayloadType)
        }

        override fun coordinatesChanged(coordinateItems: List<CoordinateItem>) {
            try {
                println("changeCoordinates:\n" + coordinateItems)

                //TODO check code updater
//                val codeUpdater = CodeUpdater()
//                javaClass.getResourceAsStream("/viewer-code-block.txt").use { resource -> println("Code:\n" + codeUpdater.updateCoordinates(IOUtils.toString(resource, StandardCharsets.UTF_8), coordinateItems)) }
            } catch (exc: Exception) {
                exc.printStackTrace()
            }

        }
    }

    override fun start(stage: Stage?) {
        if (stage == null) {
            throw IllegalArgumentException("stage is null")
        }


        val javaSourceContext = javaClass.getResource("/" + payload!!).readText()

        val parser = JavaSourceParser(object: JavaSourceParser.Listener {
            override fun error(msg: String) {
                log.error { msg }
            }
        })

        val model = parser.parse(javaSourceContext)

        val viewer = GraphViewer()
        viewer.registerListener(LogActionListener())

        viewer.setShortcut(ShortcutType.GOTO_BUILD_GRAPH, Shortcut(true, KeyCode.B))

        viewer.openGraph(model)
        stage.scene = viewer.scene

        stage.show()
    }

    /**
     * Manual UI Test
     */
    @Ignore
    @Test
    fun open_graph_for_purchase_payload() {
        payload = "viewer-test-PurchaseGraphConfig.java"
        Application.launch()
    }



}